### PR TITLE
Restore local history from translog on promotion

### PR DESCRIPTION
**Pull request by dnhatn**
If a shard was serving as a replica when another shard was promoted to
primary, then its Lucene index was reset to the global checkpoint.
However, if the new  primary fails before the primary/replica resync
completes and we are now being promoted, we have to restore the reverted
operations by replaying the translog to avoid losing acknowledged writes.

Relates #33473
Relates #32867

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -332,17 +332,12 @@ protected int getRefCount(IndexSearcher reference) {
...
             ensureOpen();
             final long localCheckpoint = localCheckpointTracker.getCheckpoint();
             try (Translog.Snapshot snapshot = getTranslog().newSnapshotFromMinSeqNo(localCheckpoint + 1)) {
-                Translog.Operation operation;
-                while ((operation = snapshot.next()) != null) {
-                    if (operation.seqNo() > localCheckpoint) {
-                        localCheckpointTracker.markSeqNoAsCompleted(operation.seqNo());
-                    }
-                }
+                return translogRecoveryRunner.run(this, snapshot);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: We can keep track a max_seqno translog to recover when we rollback this engine (i.e., recover_upto and max_seqno in translog at that time), then only restore if needed. However, I opted out for simplicity.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 2]
```diff
@@ -341,7 +341,9 @@ public void rollTranslogGeneration() {
     }
 
     @Override
-    public void restoreLocalCheckpointFromTranslog() {
+    public int restoreLocalHistoryFromTranslog(TranslogRecoveryRunner translogRecoveryRunner) {
+        assert false : "this should not be called";
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: I don't understand why this throws an exception. if you have an index that is read-only and uses this engine and a primary get's promoted this should be a no-op not a UOE?

**Review by s1monw — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: Yes, we should make this a no-op (just like fillSeqNoGaps).

**Review by s1monw — APPROVED** [review]
LGTM but I think @bleskes needs to take a look too

**Review by ywelsch — APPROVED** [review]
LGTM. Thanks @dnhatn

**Comment by dnhatn** [comment]
Thanks @s1monw and @ywelsch.

**Pull request merged**